website/flask_helpers.py

```diff
@@ -70,3 +75,84 @@ def dumps(self, obj, **kwargs):
+    'SLIDES_COM': 'https://slides.com',
+    'TEACHERS_MANUAL': 'https://www.hedy.org/for-teachers/manual',
+    'TEACHERS_MANUAL_FEATURES': 'https://hedy.org/for-teachers/manual/features',
+    'TEACHERS_MANUAL_FOR_TEACHERS': 'https://hedy.org/for-teachers/manual/preparations#for-teachers',
+    'FOR_TEACHERS': 'https://hedycode.com/for-teachers',
+    'DISCORD_SERVER': 'https://discord.gg/8yY7dEme9r',
+    'MAIL_TO_HELLO_HEDY': 'mailto: hello@hedy.org',
+    'DISCORD_VIDEO': 'https://www.youtube.com/watch?v=Lyz_Lnd-_aI',
+    'CUSTOMIZE_ADVENTURE': 'https://github.com/hedyorg/hedy/assets/80678586/df38cbb2-468e-4317-ac67-92eaf4212adc',
+}
+
+whitelisted_keys = {v for k in whitelisted_links.keys() for v in {f'#HREF_{k}#', f'#MD_{k}#', f'#SRC_{k}#'}}
+
+whitelisted_html_tags = ['a', 'b', 'em', 'strong', 'img']
+
+
+def validate_content(input_):
+    """
+    Validates that the input does not have suspicious HTML or markdown content. If there is no suspicious content,
+    the function returns True and input in which the URL references are substitute with the actual whitelisted URLs.
+    If there is suspicious content, the function returns False and the input where all special characters are escaped,
+    so that the HTML or markdown appears as plain text. For example:
+    - '<a #HREF_DISCORD_SERVER#>link</a>' returns (True, '<a href="https://discord.gg/8yY7dEme9r">link</a>')
+    - '[link](#MD_SLIDES_COM#)' returns '[link](https://slides.com)'
+    - '<a href="https://hedy.org">link</a>' is (False, '&lt;a href="https://discord.gg/8yY7dEme9r"&gt;link&lt;/a&gt;')
+    """
+    def escape_special_chars(i):
+        return (i.replace('<', '&lt;').replace('>', '&gt;')
+                .replace('[', '\\[').replace(']', '\\]')
+                .replace('(', '\\(').replace(')', '\\)'))
+
+    def replace_whitelisted_links(i):
+        for key in whitelisted_keys:
+            parts = key[1:-1].split('_')
+            value = whitelisted_links['_'.join(parts[1:])]
+            if parts[0] == 'HREF':
+                replace_by = f'href="{value}"'
+            elif parts[0] == 'SRC':
+                replace_by = f'src={value}'
+            else:
+                replace_by = value
+            i = i.replace(key, replace_by)
+        return i
+
+    def has_valid_ref(i, prefix=''):
+        href_times = i.lower().count(prefix.lower())
+        whitelist_keys_present = [link in i for link in whitelisted_keys if link.startswith(f'#{prefix.upper()}_')]
+        return href_times == 1 and any(whitelist_keys_present)
+
+    # If the content contains something which remotely reminds of an HTML or markdown comment, it is suspicious.
+    if re.search(r'<!--', input_) or re.search(r'-->', input_) or re.search(r']:', input_):
+        return False, escape_special_chars(input_)
+
+    # Find all occurrences which look like an opening HTML tag
+    matches = re.findall(r'<\s*[^/][^>]+', input_)
```

Comment:
I'm not sure about this particular regex, I've been trying it, and I'm not sure that the `[^/]` is doing anything, since it matches it anyways when it gets to the `[^>]` part

Reply:
The `[^/]` part is there so that we do not match the closing tags, which I want to disregard when looking for html content. It is still a very naive solution.